src/layer.js

```diff
@@ -0,0 +1,149 @@
+import * as util from "./util";
+import { expression } from "mapbox-gl/dist/style-spec";
+
+export default class Layer {
+  constructor({ id, source, ...options }) {
+    this.id = id;
+    this.type = "custom";
+    this.renderingMode = "2d";
+    this.source = source;
+    this._zoomUpdatable = {};
+    this._propsOnInit = {};
+
+    this.source.loadTile(0, 0, 0, this.setWind.bind(this));
+
+    Object.keys(this.propertySpec).forEach(spec => {
+      this.setProperty(spec, options[spec] || this.propertySpec[spec].default);
+    });
+  }
+
+  setProperty(prop, value) {
+    const spec = this.propertySpec[prop];
+    const expr = expression.createPropertyExpression(value, spec);
+    if (expr.result === "success") {
+      const name = prop
+        .split("-")
+        .map(a => a[0].toUpperCase() + a.slice(1))
+        .join("");
+      switch (expr.value.kind) {
+        case "camera":
+        case "composite":
+          return (this._zoomUpdatable[prop] = expr.value);
+        default:
+          if (this.map) {
+            return this._setPropertyValue(prop, expr.value);
+          } else {
+            return (this._propsOnInit[prop] = expr.value);
+          }
+      }
+    } else {
+      throw new Error(expr.value);
+    }
+  }
+
+  _setPropertyValue(prop, value) {
+    const name = prop
+      .split("-")
+      .map(a => a[0].toUpperCase() + a.slice(1))
+      .join("");
+    const setterName = "set" + name;
+    if (this[setterName]) {
+      this[setterName](value);
+    } else {
+      this[name[0].toLowerCase() + name.slice(1)] = value.evaluate({
+        zoom: this.map && this.map.getZoom()
+      });
+    }
+  }
+
+  buildColorRamp(expr) {
```

Comment:
This is a complex function, could do with docs to explain it's purpose, I'm not sure what a **Color Ramp** is.